File: backend/models/dream.py
"""
Dream Model

梦境记录数据模型。
"""

import uuid
from datetime import datetime
from typing import Optional, List

from sqlalchemy import String, Text, Integer, JSON, ForeignKey, Index
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.db.base import Base


class DreamRecord(Base):
    """梦境记录表"""
    
    __tablename__ = "dream_records"
    
    # 主键
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    
    # 用户关联
    user_id: Mapped[str] = mapped_column(
        String(64),
        nullable=False,
        index=True,
    )

    # 组织/租户ID（多租户隔离边界）
    org_id: Mapped[Optional[str]] = mapped_column(
        String(64),
        nullable=True,
        index=True,
        comment="组织/租户ID（缺省时视为 legacy 记录）",
    )
    
    # 状态和模式
    status: Mapped[str] = mapped_column(
        String(20),
        default="draft",
        nullable=False,
    )  # draft, published
    
    mode: Mapped[str] = mapped_column(
        String(20),
        default="quick",
        nullable=False,
    )  # quick, detailed
    
    # 内容
    raw_input: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    generated_content: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    
    final_content: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    # 元数据
    title: Mapped[Optional[str]] = mapped_column(
        String(200),
        nullable=True,
    )
    
    mood: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
    )
    
    clarity: Mapped[Optional[int]] = mapped_column(
        Integer,
        nullable=True,
    )  # 1-5
    
    tags: Mapped[List[str]] = mapped_column(
        ARRAY(String),
        default=list,
        nullable=False,
    )
    
    # 分析结果 (JSON)
    analysis: Mapped[Optional[dict]] = mapped_column(
        JSON,
        nullable=True,
    )
    
    # 生成计数
    generate_count: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False,
    )
    
    # 发布时间
    published_at: Mapped[Optional[datetime]] = mapped_column(
        nullable=True,
    )
    
    # 索引
    __table_args__ = (
        Index("ix_dream_records_org_user_created", "org_id", "user_id", "created_at"),
    )
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": str(self.id),
            "user_id": self.user_id,
            "org_id": self.org_id,
            "status": self.status,
            "mode": self.mode,
            "raw_input": self.raw_input,
            "generated_content": self.generated_content,
            "final_content": self.final_content,
            "title": self.title,
            "mood": self.mood,
            "clarity": self.clarity,
            "tags": self.tags,
            "analysis": self.analysis,
            "generate_count": self.generate_count,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "published_at": self.published_at.isoformat() if self.published_at else None,
        }

File: frontend/src/types/dream.ts
/**
 * Dream Journal 类型定义
 */

export type DreamMode = 'quick' | 'detailed';
export type DreamStatus = 'draft' | 'published';

export interface DreamEntry {
  id: string;
  status: DreamStatus;
  mode: DreamMode;
  
  // 内容
  rawInput: string;              // 原始输入
  generatedContent?: string;     // AI生成内容
  finalContent: string;          // 最终内容
  
  // 完整模式额外字段
  title?: string;
  date?: string;
  clarity?: number;              // 1-5星
  tags?: string[];
  mood?: string;
  
  // 元数据
  generateCount: number;         // 已使用生成次数
  createdAt: string;
  updatedAt: string;
  publishedAt?: string;
}

export interface GenerateRequest {
  keywords: string;
  mood: string;  // 必需！当时的心情（激动、幸福、紧张、害怕、崩溃等）
  userProfile?: {
    bazi?: {
      dayMaster: string;
      currentLuck: string;
    };
    lunarPhase?: string;
  };
  dreamHistory?: Array<{ pattern: string }>;
  canonicalSymbols?: Record<string, string>;
}

export interface GenerateResponse {
  success: boolean;
  narrative?: string;
  tokensUsed?: number;
  model?: string;
  error?: string;
}

export interface UserLimits {
  dailyDreamsRemaining: number;  // 今日剩余梦境数（最多3个）
  totalDreamsToday: number;      // 今日已记录数
}
